extensions/amp-access/0.1/amp-access.js

```diff
@@ -233,6 +256,108 @@ export class AccessService {
   }
 
   /**
+   * @private
+   */
+  scheduleView_() {
+    if (this.viewer_.isVisible()) {
+      this.reportWhenViewed_();
+    }
+    this.viewer_.onVisibilityChanged(() => {
+      if (this.viewer_.isVisible()) {
+        this.reportWhenViewed_();
+      }
+    });
+  }
+
+  /**
+   * @return {!Promise}
+   * @private
+   */
+  reportWhenViewed_() {
+    if (this.reportViewPromise_) {
+      return this.reportViewPromise_;
+    }
+    log.fine(TAG, 'start view monitoring');
+    this.reportViewPromise_ = this.whenViewed_().then(
+        this.reportViewToServer_.bind(this),
+        reason => {
+          // Ignore - view has been canceled.
+          log.fine(TAG, 'view cancelled:', reason);
+          this.reportViewPromise_ = undefined;
+          throw reason;
+        });
+    return this.reportViewPromise_;
+  }
+
+  /**
+   * The promise will be resolved when a view of this document has occurred. It
+   * will be rejected if the current impression should not be counted as a view.
+   * @return {!Promise}
+   * @private
+   */
+  whenViewed_() {
+    // Viewing kick off: document is visible.
+    const unlistenSet = [];
+
+    // Once document is visible we have a competing process for triggering
+    // condition and when the document becomes invisible again.
+    let resolve, reject;
+    const promise = new Promise((aResolve, aReject) => {
+      resolve = aResolve;
+      reject = aReject;
+    });
+
+    // 1. Document becomes invisible again: cancel.
+    unlistenSet.push(this.viewer_.onVisibilityChanged(() => {
+      if (!this.viewer_.isVisible()) {
+        reject();
+      }
+    }));
+
+    // 2. After a few seconds: register a view.
+    this.docState_.onReady(() => {
+      timer.delay(resolve, VIEW_TIMEOUT);
+    });
+
+    // 3. If scrolled: register a view.
+    unlistenSet.push(this.viewport_.onScroll(resolve));
+
+    // 4. Tap: register a view.
+    unlistenSet.push(listenOnce(this.win.document.documentElement,
+        'click', resolve));
+
+    return promise.then(() => {
+      unlistenSet.forEach(unlisten => unlisten());
+    }, reason => {
+      unlistenSet.forEach(unlisten => unlisten());
+      throw reason;
+    });
+  }
+
+  /**
+   * @return {!Promise}
+   * @private
+   */
+  reportViewToServer_() {
+    return this.buildUrl_(this.config_.pingback).then(url => {
+      log.fine(TAG, 'Pingback URL: ', url);
+      return this.xhr_.sendSignal(url, {
+        method: 'POST',
+        credentials: 'include',
+        headers: {
+          'Content-Type': 'application/x-www-form-urlencoded'
+        },
+        body: ''
+      });
+    }).then(() => {
+      log.fine(TAG, 'Pingback complete');
+    }).catch(error => {
+      log.error(TAG, 'Pingback failed: ', error);
+      throw error;
```

Comment:
I'm exposing these in expectation that these errors will be substituted by publisher logs, in which case I think they'd be useful. But, of course, i'll be happy to remove if you feel like it's not the case.